multichain-aggregator/multichain-aggregator-logic/src/repository/addresses.rs

```diff
@@ -75,21 +68,112 @@ where
                 .to_owned(),
         )
         .materialized(is_cte_materialized)
-        .table_name(addresses_cte_iden.clone())
+        .table_name(name)
+        .to_owned()
+}
+
+pub async fn uniform_chain_search<C>(
+    db: &C,
+    query: String,
+    token_types: Option<Vec<db_enum::TokenType>>,
+    chain_ids: Vec<ChainId>,
+) -> Result<Vec<Model>, DbErr>
+where
+    C: ConnectionTrait,
+{
+    if chain_ids.is_empty() {
+        return Ok(vec![]);
+    }
+
+    let addresses_cte_iden = Alias::new("addresses").into_iden();
+    let addresses_cte = prepare_address_search_cte(Some(query), addresses_cte_iden.clone());
+
+    let row_number = Expr::custom_keyword(Alias::new("ROW_NUMBER()"));
+    let ranked_addresses_iden = Alias::new("ranked_addresses").into_iden();
+    let ranked_addresses_cte = CommonTableExpression::new()
+        .query(
+            Query::select()
+                .column(ColumnRef::TableAsterisk(addresses_cte_iden.clone()))
+                .expr_window_as(
+                    row_number,
+                    WindowStatement::partition_by(Column::ChainId),
+                    Alias::new("rn"),
+                )
+                .from(addresses_cte_iden.clone())
+                .and_where(Column::ChainId.is_in(chain_ids.clone()))
+                .apply_if(token_types, |q, token_types| {
+                    if !token_types.is_empty() {
+                        q.and_where(Column::TokenType.is_in(token_types));
+                    } else {
+                        q.and_where(Column::TokenType.is_null());
+                    }
+                })
+                .to_owned(),
+        )
+        .table_name(ranked_addresses_iden.clone())
+        .to_owned();
+
+    let limit = chain_ids.len() as u64;
+    let base_select = Query::select()
+        .column(ColumnRef::Asterisk)
+        .from(ranked_addresses_iden)
+        .and_where(Expr::col(Alias::new("rn")).eq(1))
```

Comment:
is there any determenistic ordering within groups from `PARTITION BY chain_id`? it seems that there should be some rank that will make `rn=1` entries to be the best matching, but I don't see `ORDER BY` that does this.

Maybe I missed something?

Reply:
from what I understand, `ORDER BY array_position($4, chain_id) DESC` only sorts the 'groups' themselves

Reply:
> is there any determenistic ordering within groups from PARTITION BY chain_id?

There is no ordering for now. I guess we can add at least some `ts_rank` + `primary_key` order to make deterministic and more looking like a best match search.

>from what I understand, ORDER BY array_position($4, chain_id) DESC only sorts the 'groups' themselves

That is true. This is needed only to sort the chains according to their priority.